sys/net/gnrc/conn/tcp/gnrc_conn_tcp.c

```diff
@@ -0,0 +1,560 @@
+        return -EBADF;
+    }
+
+    /* Guard: Connection is in an illegal state for this operation */
+    if (conn->state != ESTABLISHED && conn->state != CLOSE_WAIT) {
+        return -EALREADY;
+    }
+
+    /* Purge message queue */
+    while(msg_try_receive(&msg) >= 0){
+    }
+
+    /* Setup User Space Timeout */
+    msg_tout.type = MSG_TYPE_USER_TIMEOUT;
+    xtimer_set_msg(&usr_tout, GNRC_TCP_USER_TIMEOUT_DURATION, &msg_tout, thread_getpid());
+
+    /* Loop until something was sent and acked */
+    while (err == 0 || conn->pkt_retransmit != NULL) {
+
+        /* If the send window is closed: Setup Probing */
+        if (conn->snd_wnd <= 0) {
+            /* If this is the first probe: Setup probing duration */
+            if (!probing) {
+                probing = true;
+                tout_probe_duration = conn->rto;
+            }
+            /* Initialize Probe Timer */
+            msg_probe.type = MSG_TYPE_PROBE_TIMEOUT;
+            xtimer_set_msg(&tout_probe, tout_probe_duration, &msg_probe, thread_getpid());
+        }
+
+        /* Try to send data in case there nothing has been sent and we are not probing */
+        if (err == 0 && !probing) {
+            err = _fsm(conn, CALL_SEND, NULL, (void *) data, len);
+        }
+        /* Wait for responses */
+        msg_receive(&msg);
+        switch (msg.type) {
+            case MSG_TYPE_USER_TIMEOUT:
+                DEBUG("gnrc_conn_tcp : conn_tcp_send() : USER_TIMEOUT\n");
+                if (probing) {
+                    xtimer_remove(&tout_probe);
+                }
+                _fsm(conn, TIMEOUT_USER, NULL, NULL, 0);
+                err = -ETIME;
+                break;
+
+            case MSG_TYPE_PROBE_TIMEOUT:
+                DEBUG("gnrc_conn_tcp : conn_tcp_send() : PROBE_TIMEOUT\n");
+
+                /* Send Probe */
+                _fsm(conn, SEND_PROBE, NULL, NULL, 0);
+                tout_probe_duration += tout_probe_duration;
+
+                /* Boundry check for time interval between probes */
+                if (tout_probe_duration < GNRC_TCP_PROBE_LOWER_BOUND) {
+                    tout_probe_duration = GNRC_TCP_PROBE_LOWER_BOUND;
+                }
+                else if (tout_probe_duration > GNRC_TCP_PROBE_UPPER_BOUND) {
+                    tout_probe_duration = GNRC_TCP_PROBE_UPPER_BOUND;
+                }
+                break;
+
+            case MSG_TYPE_NOTIFY_USER:
+                DEBUG("gnrc_conn_tcp : conn_tcp_send() : MSG_TYPE_NOTIFY_USER\n");
+                /* Connection is alive: Reset User Space Timeout */
+                xtimer_set_msg(&usr_tout, GNRC_TCP_USER_TIMEOUT_DURATION, &msg_tout, thread_getpid());
+
+                /* If the window re-opened and we are probing: Stop it */
+                if (conn->snd_wnd > 0 && probing) {
+                    probing = false;
+                    xtimer_remove(&tout_probe);
+                }
+                break;
```

Comment:
The logic of `MSG_TYPE_PROBE_TIMEOUT` and `MSG_TYPE_NOTIFY_USER` could also be implemented in the fsm.